src/uct/tcp/tcp_ep.c

```diff
@@ -62,64 +74,211 @@ static void uct_tcp_ep_epoll_ctl(uct_tcp_ep_t *ep, int op)
+    }
+
+    memset(sock_addr, 0, sizeof(*sock_addr));
+}
+
+static ucs_status_t uct_tcp_ep_addr_init(ucs_sock_addr_t *sock_addr,
+                                         const struct sockaddr *addr)
+{
+    socklen_t addr_len = sizeof(struct sockaddr_in); /* TODO */
+    struct sockaddr *new_addr;
+
+    if (sock_addr == NULL) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    if (addr == NULL) {
+        uct_tcp_ep_addr_cleanup(sock_addr);
+    } else {
+        new_addr = ucs_malloc(addr_len, "sock_addr");
+        if (new_addr == NULL) {
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        sock_addr->addr    = new_addr;
+        sock_addr->addrlen = addr_len;
+    }
+
+    return UCS_OK;
+}
+
+static void uct_tcp_ep_close_fd(int *fd_p)
+{
+    if ((fd_p != NULL) && (*fd_p != -1)) {
+        close(*fd_p);
+    }
+}
+
+static unsigned uct_tcp_ep_in_iface(uct_tcp_ep_t *ep)
+{
+    return !ucs_list_is_empty(&ep->list);
+}
+
+static void uct_tcp_ep_del_from_iface(uct_tcp_iface_t *iface,
+                                      uct_tcp_ep_t *ep)
+{
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    if (uct_tcp_ep_in_iface(ep)) {
+        ucs_list_del(&ep->list);
+    }
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static void uct_tcp_ep_add_to_iface(uct_tcp_iface_t *iface,
+                                    uct_tcp_ep_t *ep)
+{
+    UCS_ASYNC_BLOCK(iface->super.worker->async);
+    ucs_list_add_tail(&iface->ep_list, &ep->list);
+    UCS_ASYNC_UNBLOCK(iface->super.worker->async);
+}
+
+static void uct_tcp_ep_cleanup(uct_tcp_ep_t *ep)
+{ 
+    uct_tcp_ep_addr_cleanup(&ep->peer_addr);
+
+    uct_tcp_ep_ctx_cleanup(&ep->tx);
+    uct_tcp_ep_ctx_cleanup(&ep->rx);
+
+    uct_tcp_ep_close_fd(&ep->fd);
+}
+
+static ucs_status_t uct_tcp_ep_init(uct_tcp_iface_t *iface,
+                                    uct_tcp_ep_t *ep,
+                                    const struct sockaddr *dest_addr)
+{
+    ucs_status_t status;
+
+    ep->peer_addr.addr = NULL;
+
+    status = uct_tcp_ep_addr_init(&ep->peer_addr, dest_addr);
+    if (status != UCS_OK) {
+        return UCS_OK;
+    }
+
+    ep->tx.buf = NULL;
+    ep->tx.progress = uct_tcp_ep_empty_progress;
+
+    ep->rx.buf = NULL;
+    ep->rx.progress = uct_tcp_ep_empty_progress;
+
+    ucs_queue_head_init(&ep->pending_q);
+    ep->events = 0;
+    ep->fd = -1;
+    ucs_list_head_init(&ep->list);
+
+    return UCS_OK;
 }
 
 static UCS_CLASS_INIT_FUNC(uct_tcp_ep_t, uct_tcp_iface_t *iface,
-                           int fd, const struct sockaddr_in *dest_addr)
+                           int fd, const struct sockaddr *dest_addr)
 {
     ucs_status_t status;
 
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super)
 
-    self->buf = ucs_malloc(ucs_max(iface->config.buf_size,
-                                   iface->config.short_size), "tcp_buf");
-    if (self->buf == NULL) {
-        return UCS_ERR_NO_MEMORY;
+    status = uct_tcp_ep_init(iface, self, dest_addr);
```

Comment:
why need another function?
it's called only once and it's not that big

also, looks like fd==-1 always now, so the `if (fd == -1)` in line 238 is not needed any more?